File: app/src/main/java/wangdaye/com/geometricweather/main/ui/adapter/AqiAdapter.java
package wangdaye.com.geometricweather.main.ui.adapter;

import android.animation.AnimatorSet;
import android.animation.ArgbEvaluator;
import android.animation.FloatEvaluator;
import android.animation.ValueAnimator;
import android.content.Context;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.ColorUtils;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.DecelerateInterpolator;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.basic.model.weather.Weather;
import wangdaye.com.geometricweather.main.ui.MainColorPicker;
import wangdaye.com.geometricweather.ui.widget.RoundProgress;
import wangdaye.com.geometricweather.weather.WeatherHelper;

/**
 * Aqi adapter.
 */

public class AqiAdapter extends RecyclerView.Adapter<AqiAdapter.ViewHolder> {

    private List<AqiItem> itemList;
    private List<ViewHolder> holderList;
    private MainColorPicker colorPicker;

    private class AqiItem {
        @ColorInt int color;
        float progress;
        float max;
        String title;
        String content;

        boolean executeAnimation;

        AqiItem(@ColorInt int color, float progress, float max, String title, String content,
                boolean executeAnimation) {
            this.color = color;
            this.progress = progress;
            this.max = max;
            this.title = title;
            this.content = content;
            this.executeAnimation = executeAnimation;
        }
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        @Nullable AqiItem item;
        private boolean executeAnimation;
        @Nullable private AnimatorSet attachAnimatorSet;

        private TextView title;
        private TextView content;
        private RoundProgress progress;

        ViewHolder(View itemView) {
            super(itemView);
            this.title = itemView.findViewById(R.id.item_aqi_title);
            this.content = itemView.findViewById(R.id.item_aqi_content);
            this.progress = itemView.findViewById(R.id.item_aqi_progress);
        }

        void onBindView(AqiItem item) {
            Context context = itemView.getContext();

            this.item = item;
            this.executeAnimation = item.executeAnimation;

            title.setText(item.title);
            title.setTextColor(colorPicker.getTextContentColor(context));

            content.setText(item.content);
            title.setTextColor(colorPicker.getTextSubtitleColor(context));

            if (executeAnimation) {
                progress.setProgress(0);
                progress.setProgressColor(ContextCompat.getColor(context, R.color.colorLevel_1));
                progress.setProgressBackgroundColor(colorPicker.getLineColor(context));
            } else {
                progress.setProgress((int) (100.0 * item.progress / item.max));
                progress.setProgressColor(item.color);
                progress.setProgressBackgroundColor(
                        ColorUtils.setAlphaComponent(item.color, (int) (255 * 0.1))
                );
            }
        }

        void executeAnimation() {
            if (executeAnimation && item != null) {
                executeAnimation = false;

                ValueAnimator progressColor = ValueAnimator.ofObject(
                        new ArgbEvaluator(),
                        ContextCompat.getColor(itemView.getContext(), R.color.colorLevel_1),
                        item.color
                );
                progressColor.addUpdateListener(animation ->
                        progress.setProgressColor((Integer) animation.getAnimatedValue())
                );

                ValueAnimator backgroundColor = ValueAnimator.ofObject(
                        new ArgbEvaluator(),
                        colorPicker.getLineColor(itemView.getContext()),
                        ColorUtils.setAlphaComponent(item.color, (int) (255 * 0.1))
                );
                backgroundColor.addUpdateListener(animation ->
                        progress.setProgressBackgroundColor((Integer) animation.getAnimatedValue())
                );

                ValueAnimator aqiNumber = ValueAnimator.ofObject(new FloatEvaluator(), 0, item.progress);
                aqiNumber.addUpdateListener(animation ->
                        progress.setProgress(
                                100.0f * ((Float) animation.getAnimatedValue()) / item.max
                        )
                );

                attachAnimatorSet = new AnimatorSet();
                attachAnimatorSet.playTogether(progressColor, backgroundColor, aqiNumber);
                attachAnimatorSet.setInterpolator(new DecelerateInterpolator(3));
                attachAnimatorSet.setDuration((long) (item.progress / item.max * 5000));
                attachAnimatorSet.start();
            }
        }

        void cancelAnimation() {
            if (attachAnimatorSet != null && attachAnimatorSet.isRunning()) {
                attachAnimatorSet.cancel();
            }
            attachAnimatorSet = null;
        }
    }

    public AqiAdapter(Context context, Weather weather,
                      MainColorPicker colorPicker, boolean executeAnimation) {
        this.itemList = new ArrayList<>();
        if (weather != null && weather.aqi != null) {
            if (weather.aqi.pm25 >= 0) {
                itemList.add(
                        new AqiItem(
                                WeatherHelper.getPm25Color(context, weather.aqi.pm25),
                                weather.aqi.pm25,
                                250,
                                "PM2.5",
                                weather.aqi.pm25 + " μg/m³",
                                executeAnimation
                        )
                );
            }
            if (weather.aqi.pm10 >= 0) {
                itemList.add(
                        new AqiItem(
                                WeatherHelper.getPm10Color(context, weather.aqi.pm10),
                                weather.aqi.pm10,
                                420,
                                "PM10",
                                weather.aqi.pm10 + " μg/m³",
                                executeAnimation
                        )
                );
            }
            if (weather.aqi.so2 >= 0) {
                itemList.add(
                        new AqiItem(
                                WeatherHelper.getSo2Color(context, weather.aqi.so2),
                                weather.aqi.so2,
                                1600,
                                "SO2",
                                weather.aqi.so2 + " μg/m³",
                                executeAnimation
                        )
                );
            }
            if (weather.aqi.no2 >= 0) {
                itemList.add(
                        new AqiItem(
                                WeatherHelper.getNo2Color(context, weather.aqi.no2),
                                weather.aqi.no2,
                                565,
                                "NO2",
                                weather.aqi.no2 + " μg/m³",
                                executeAnimation
                        )
                );
            }
            if (weather.aqi.o3 >= 0) {
                itemList.add(
                        new AqiItem(
                                WeatherHelper.getO3Color(context, weather.aqi.o3),
                                weather.aqi.o3,
                                800,
                                "O3",
                                weather.aqi.o3 + " μg/m³",
                                executeAnimation
                        )
                );
            }
            if (weather.aqi.co >= 0) {
                itemList.add(
                        new AqiItem(
                                WeatherHelper.getCOColor(context, weather.aqi.co),
                                weather.aqi.co,
                                90,
                                "CO",
                                weather.aqi.co + " μg/m³",
                                executeAnimation
                        )
                );
            }
        }

        this.holderList = new ArrayList<>();
        this.colorPicker = colorPicker;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_aqi, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.onBindView(itemList.get(position));
        if (itemList.get(position).executeAnimation) {
            holderList.add(holder);
        }
    }

    @Override
    public int getItemCount() {
        return itemList.size();
    }

    public void executeAnimation() {
        for (int i = 0; i < holderList.size(); i++) {
            holderList.get(i).executeAnimation();
        }
    }

    public void cancelAnimation() {
        for (int i = 0; i < holderList.size(); i++) {
            holderList.get(i).cancelAnimation();
        }
        holderList.clear();
    }
}

File: app/src/main/java/org/breezyweather/ui/main/adapters/AqiAdapter.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.ui.main.adapters

import android.animation.AnimatorSet
import android.animation.ArgbEvaluator
import android.animation.FloatEvaluator
import android.animation.ValueAnimator
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.DecelerateInterpolator
import android.widget.TextView
import androidx.annotation.ColorInt
import androidx.core.content.ContextCompat
import androidx.core.graphics.ColorUtils
import androidx.recyclerview.widget.RecyclerView
import breezyweather.domain.location.model.Location
import kotlinx.collections.immutable.ImmutableList
import kotlinx.collections.immutable.toImmutableList
import org.breezyweather.R
import org.breezyweather.domain.weather.index.PollutantIndex
import org.breezyweather.domain.weather.model.getColor
import org.breezyweather.domain.weather.model.getConcentration
import org.breezyweather.domain.weather.model.getIndex
import org.breezyweather.domain.weather.model.validAirQuality
import org.breezyweather.ui.common.widgets.RoundProgress
import org.breezyweather.ui.main.utils.MainThemeColorProvider

class AqiAdapter(
    context: Context,
    location: Location,
    executeAnimation: Boolean,
) : RecyclerView.Adapter<AqiAdapter.ViewHolder>() {
    private val mLightTheme: Boolean
    private val mItemList: ImmutableList<AqiItem>
    private val mHolderList: MutableList<ViewHolder>

    class AqiItem(
        val pollutantType: PollutantIndex,
        @field:ColorInt val color: Int,
        val progress: Float,
        val max: Float,
        val title: String,
        val content: String,
        val talkBack: String,
        val executeAnimation: Boolean,
    )

    class ViewHolder(
        itemView: View,
    ) : RecyclerView.ViewHolder(itemView) {
        private var mItem: AqiItem? = null
        private var mLightTheme: Boolean? = null
        private var mExecuteAnimation = false
        private var mAttachAnimatorSet: AnimatorSet? = null
        private val mTitle: TextView = itemView.findViewById(R.id.item_aqi_title)
        private val mContent: TextView = itemView.findViewById(R.id.item_aqi_content)
        private val mProgress: RoundProgress = itemView.findViewById(R.id.item_aqi_progress)

        fun onBindView(lightTheme: Boolean, item: AqiItem) {
            val context = itemView.context
            mItem = item
            mLightTheme = lightTheme
            mExecuteAnimation = item.executeAnimation
            itemView.contentDescription = item.talkBack
            mTitle.text = item.title
            mTitle.setTextColor(MainThemeColorProvider.getColor(lightTheme, R.attr.colorTitleText))
            mContent.text = item.content
            mContent.setTextColor(MainThemeColorProvider.getColor(lightTheme, R.attr.colorBodyText))
            itemView.setOnClickListener {
                // TODO: Open daily details activity on air quality page
                /*IntentHelper.startDailyWeatherActivity(
                    context as GeoActivity,
                    location.formattedId,
                    location.weather!!.todayIndex,
                    ChartDisplay.TAG_AIR_QUALITY
                )*/
            }
            if (mExecuteAnimation) {
                mProgress.apply {
                    progress = 0f
                    setProgressColor(ContextCompat.getColor(context, R.color.colorLevel_1))
                    setProgressBackgroundColor(
                        MainThemeColorProvider.getColor(lightTheme, com.google.android.material.R.attr.colorOutline)
                    )
                }
            } else {
                mProgress.apply {
                    progress = (100.0 * item.progress / item.max).toInt().toFloat()
                    setProgressColor(item.color)
                    setProgressBackgroundColor(
                        ColorUtils.setAlphaComponent(item.color, (255 * 0.1).toInt())
                    )
                }
            }
        }

        fun executeAnimation() {
            if (mExecuteAnimation) {
                mItem?.let { item ->
                    mExecuteAnimation = false
                    val progressColor = ValueAnimator.ofObject(
                        ArgbEvaluator(),
                        ContextCompat.getColor(itemView.context, R.color.colorLevel_1),
                        item.color
                    )
                    progressColor.addUpdateListener { animation: ValueAnimator ->
                        mProgress.setProgressColor((animation.animatedValue as Int))
                    }
                    val backgroundColor = ValueAnimator.ofObject(
                        ArgbEvaluator(),
                        MainThemeColorProvider.getColor(
                            mLightTheme ?: false,
                            com.google.android.material.R.attr.colorOutline
                        ),
                        ColorUtils.setAlphaComponent(item.color, (255 * 0.1).toInt())
                    )
                    backgroundColor.addUpdateListener { animation: ValueAnimator ->
                        mProgress.setProgressBackgroundColor((animation.animatedValue as Int))
                    }
                    val aqiNumber = ValueAnimator.ofObject(FloatEvaluator(), 0, item.progress)
                    aqiNumber.addUpdateListener { animation: ValueAnimator ->
                        mProgress.progress = 100.0f * animation.animatedValue as Float / item.max
                    }
                    mAttachAnimatorSet = AnimatorSet().apply {
                        playTogether(progressColor, backgroundColor, aqiNumber)
                        interpolator = DecelerateInterpolator(3f)
                        duration = (item.progress / item.max * 5000).toLong()
                        start()
                    }
                }
            }
        }

        fun cancelAnimation() {
            mAttachAnimatorSet?.let {
                if (it.isRunning) it.cancel()
            }
            mAttachAnimatorSet = null
        }
    }

    init {
        mLightTheme = MainThemeColorProvider.isLightTheme(context, location)
        mItemList = buildList {
            location.weather?.validAirQuality?.let { airQuality ->
                // We use air quality index for the progress bar instead of concentration for more realistic bar
                // We use air quality index for the progress bar instead of concentration for more realistic bar
                listOf(PollutantIndex.PM25, PollutantIndex.PM10, PollutantIndex.O3, PollutantIndex.NO2)
                    .forEach { pollutantIndex ->
                        airQuality.getConcentration(pollutantIndex)?.let {
                            add(
                                AqiItem(
                                    pollutantIndex,
                                    airQuality.getColor(context, pollutantIndex),
                                    airQuality.getIndex(pollutantIndex)!!.toFloat(),
                                    PollutantIndex.indexExcessivePollution.toFloat(),
                                    context.getString(pollutantIndex.shortName),
                                    PollutantIndex.getUnit(pollutantIndex).getValueText(context, it),
                                    context.getString(pollutantIndex.voicedName) +
                                        context.getString(R.string.colon_separator) +
                                        PollutantIndex.getUnit(pollutantIndex).getValueVoice(context, it),
                                    executeAnimation
                                )
                            )
                        }
                    }
                listOf(PollutantIndex.SO2, PollutantIndex.CO)
                    .forEach { pollutantIndex ->
                        (airQuality.getConcentration(pollutantIndex) ?: 0.0).let {
                            if (it > 0) {
                                add(
                                    AqiItem(
                                        pollutantIndex,
                                        airQuality.getColor(context, pollutantIndex),
                                        airQuality.getIndex(pollutantIndex)!!.toFloat(),
                                        PollutantIndex.indexExcessivePollution.toFloat(),
                                        context.getString(pollutantIndex.shortName),
                                        PollutantIndex.getUnit(pollutantIndex).getValueText(context, it),
                                        context.getString(pollutantIndex.voicedName) +
                                            context.getString(R.string.colon_separator) +
                                            PollutantIndex.getUnit(pollutantIndex).getValueVoice(context, it),
                                        executeAnimation
                                    )
                                )
                            }
                        }
                    }
            }
        }.toImmutableList()
        mHolderList = mutableListOf()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_aqi, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.onBindView(mLightTheme, mItemList[position])
        if (mItemList[position].executeAnimation) {
            mHolderList.add(holder)
        }
    }

    override fun getItemCount(): Int {
        return mItemList.size
    }

    fun executeAnimation() {
        for (viewHolder in mHolderList) {
            viewHolder.executeAnimation()
        }
    }

    fun cancelAnimation() {
        for (viewHolder in mHolderList) {
            viewHolder.cancelAnimation()
        }
        mHolderList.clear()
    }
}
